staging/src/k8s.io/client-go/tools/cache/the_real_fifo.go

```diff
@@ -57,11 +65,23 @@ var _ QueueWithBatch = &RealFIFO{}
 // 1. delivers notifications for items that have been deleted
 // 2. delivers multiple notifications per item instead of simply the most recent value
 type RealFIFO struct {
+	// logger is a per-instance logger. This gets chosen when constructing
+	// the instance, with klog.Background() as default.
+	logger klog.Logger
+
+	// name is the name of the fifo. It is included in the logger.
+	name string
+
 	lock sync.RWMutex
 	cond sync.Cond
 
 	items []Delta
 
+	// synced is initially an open channel. It gets closed (once!) by hasSynced_locked
```

Comment:
Only the initial sync? What about cases where we replace the underlying queue, previously this would be considered unsynced until the initial population is processed but now the channel will be closed after whatever the initial sync does. This differs from the hasSynced behavior, just want to make sure we think this through before expanding the interface. Should replace also swap out the channel?

Reply:
> Only the initial sync?

The usual semantic of `HasSynced` is that once an entity is synced, it remains synced. Callers which use `WaitForCacheSync` proceed as soon as they observe "true" from all entities.

I agree that this is a change of behavior, but IMHO it makes it more consistent.

> What about cases where we replace the underlying queue

Is that something that is being done?

> Should replace also swap out the channel?

Maybe, but then we have the same race as with `HasSynced` flipping back to `false` from `true`: some client waiting might have moved on already.